Networking/Networking/CookieNonce/CookieNonceAuthenticator.swift

```diff
@@ -0,0 +1,169 @@
+import Alamofire
+import Foundation
+
+final class CookieNonceAuthenticator: RequestRetrier & RequestAdapter {
+    private let username: String
+    private let password: String
+    private let loginURL: URL
+    private let adminURL: URL
+    private var nonce: String?
+    // If we can't get this to work once, don't retry for the same site
+    // It is likely that there is something preventing us from extracting a nonce
+    private var canRetry = true
+    private var isAuthenticating = false
+    private var requestsToRetry = [RequestRetryCompletion]()
+
+    init(configuration: CookieNonceAuthenticatorConfiguration) {
+        self.username = configuration.username
+        self.password = configuration.password
+        self.loginURL = configuration.loginURL
+        self.adminURL = configuration.adminURL
+    }
+
+    // MARK: Request Adapter
+
+    func adapt(_ urlRequest: URLRequest) throws -> URLRequest {
+        guard let nonce = nonce else {
+            return urlRequest
+        }
+        var adaptedRequest = urlRequest
+        adaptedRequest.addValue(nonce, forHTTPHeaderField: "X-WP-Nonce")
+        return adaptedRequest
+    }
+
+    // MARK: Retrier
+    func should(_ manager: SessionManager, retry request: Alamofire.Request, with error: Swift.Error, completion: @escaping RequestRetryCompletion) {
+        guard
+            canRetry,
+            // Only retry once
+            request.retryCount == 0,
+            // And don't retry the login request
+            request.request?.url != loginURL,
+            // Only retry because of failed authorization
+            case .responseValidationFailed(reason: .unacceptableStatusCode(code: 401)) = error as? AFError
+        else {
+            return completion(false, 0.0)
+        }
+
+        requestsToRetry.append(completion)
+        if !isAuthenticating {
+            startLoginSequence(manager: manager)
+        }
+    }
+
+    enum Error: Swift.Error {
+        case invalidNewPostURL
+        case postLoginFailed(Swift.Error)
+        case missingNonce
+        case unknown(Swift.Error)
+    }
+}
+
+// MARK: Private helpers
+private extension CookieNonceAuthenticator {
+
+    func startLoginSequence(manager: SessionManager) {
+        DDLogInfo("Starting Cookie+Nonce login sequence for \(loginURL)")
+        guard let nonceRetrievalURL = buildNonceRequestURL(base: adminURL),
+              let nonceRequest = try? URLRequest(url: nonceRetrievalURL, method: .get) else {
+            return invalidateLoginSequence(error: .invalidNewPostURL)
+        }
+        Task(priority: .medium) {
```

Comment:
I think for networking requests, we want to use priorities lower than `.userInitiated` to avoid blocking the UI. `.medium` is my go-to for higher-priority networking requests - I hope that makes sense.